Triton dequantization/config framework

blepping opened the pull request:
Continuation of #331. Sorry for making a new pull, I couldn't reopen the existing one since it had been force pushed.

Wrapping the existing dequant functions with `torch.compile` in my benchmark tool brought the results to parity with Triton: https://gist.github.com/blepping/963459b244a4140b081cebdec24c56b2

However, I was not able to replicate that in the real world so the Triton kernels are back!

This includes a generic approach to pass configuration parameters around, including overriding the dequant functions (potentially with compiled or Triton versions). There is an `optimize` parameter in the advanced loader that lets you choose between `none`, `compile` and `triton`. Enabling Triton for me is noticeably faster, using `torch.compile` is actually slower than nothing.

This also fixes an issue with the existing approach to setting configuration like `dequant_dtype`:

```python
        ops = GGMLOps()

        if dequant_dtype in ("default", None):
            ops.Linear.dequant_dtype = None
        elif dequant_dtype in ["target"]:
            ops.Linear.dequant_dtype = dequant_dtype
        else:
            ops.Linear.dequant_dtype = getattr(torch, dequant_dtype)
```

This is not kosher because while `ops` is an instance of `GGMLOps`, attributes like `ops.Linear` are **not** instance specific, so setting it here will change that attribute in the global class. This means if there are multiple loaders with different settings, you will get the configuration of whatever loader overwrote that value last.

blepping committed:
Experimental Triton support for Q8_0 and Q4_K

blepping committed:
Add Q6_K Triton kernel

blepping committed:
Add Q5_K Triton kernel

blepping committed:
Actually enable the Q5_K kernel

blepping committed:
Triton dequant code cleanups

blepping committed:
Use static_range in Triton kernels

blepping committed:
Refactor Q4_K Triton kernel a bit to reduce code duplication

blepping committed:
Refactor/cleanup Triton support
Remove Q8_0 Triton kernel

blepping committed:
Fix dequant_dtype handling
Fix module config handling
Allow compiling GGUF dequant functions

blepping committed:
Add an optimize parameter to the advanced loader

blepping commented:
<img width="817" height="566" alt="image" src="https://github.com/user-attachments/assets/46d3b367-1f35-4c96-8505-bbf519ca3aa1" />

Some helpful test feedback I got on Discord. The Triton performance benefit is about around what I've observed in my own testing. Interesting to see that it's pretty constant even across Nvidia and AMD.

StrongerXi commented:
Hmm I think if you use the comfyui builtin or Kijai's torch compile nodes, you should get the same, or maybe even better perf than the triton kernels. What torch.compile does under the hood is generating optimized triton kernels automatically.

patientx commented:
> Hmm I think if you use the comfyui builtin or Kijai's torch compile nodes, you should get the same, or maybe even better perf than the triton kernels. What torch.compile does under the hood is generating optimized triton kernels automatically. 

with wan this doesn't make a difference , tried with q4km BUT as you can see from my tests it certainly works. BTW , I am also using the standard torch.compile and sage-attention on top of this so this is just about gguf itself and it works.

blepping committed:
Implement Q3_K Triton kernel
Refactoring/cleanups

blepping committed:
Remove unnecessary to_uint32 helper function in dequant.py

blepping committed:
Implement Q2_K, Q4_0, Q4_1, Q5_0 and Q5_1 Triton kernels
Refactoring/cleanups

blepping commented:
Q8_0 isn't worth it, but as of now there are now Triton kernels for all the K-quants and legacy quants.

The following tests were done using this utility I made: https://gist.github.com/blepping/963459b244a4140b081cebdec24c56b2

Using it requires running from the repo directory with this branch checkout and a venv with the necessary dependencies (PyTorch, Triton, etc) activated.

I used this command (Linux/Unix, presumably would work with WSL) to run it over all the available quants and dtypes: `for dt in float32 float16 bfloat16; do for qt in q4_0 q4_1 q5_0 q5_1 q2_k q3_k q4_k q5_k q6_k; do python -O test_triton_dequant.py --no-validate --qtype=$qt --dtype=$dt; done; done`

## PyTorch vs Triton

Comparing performance of the Triton kernels vs the existing PyTorch dequant functions. 2.0 in a column would mean the Triton version was two times faster. These results are from benchmarking the dequant functions in isolation so you won't see the same speedup running an actual model.

For reference, Q4_K is ~3.5x here, for moderate image sizes with models like Flux, Qwen the real world performance benefit is more like 1.2x. The Q8_0 kernel which wasn't worth using was around 1.4x here I will have to do some real testing with the quants that seem a bit borderline to find out if having them enabled is actually worth it (Q4_0, Q2_K at non-32bit, etc).

| qtype | float32 | float16 | bfloat16 |
| - | - | - | - |
| Q4_0 | 2.39 | 2.41 | 2.37 |
| Q4_1 | 3.07 | 2.42 | 2.39 |
| Q5_0 | 5.55 | 5.75 | 5.67 |
| Q5_1 | 6.14 | 5.72 | 5.45 |
| Q2_K | 3.61 | 2.52 | 2.57 |
| Q3_K | 3.47 | 3.29 | 3.17 |
| Q4_K | 3.54 | 3.91 | 3.75 |
| Q5_K | 4.64 | 4.61 | 4.67 |
| Q6_K | 3.82 | 4.13 | 4.29 |

***

## PyTorch vs PyTorch float32

Comparing the PyTorch dequant functions with `float16` and `bfloat16` vs `float32`. For example 1.27 would mean dequantizing for that dtype was 1.27x the performance of dequantizing at `float32`. _Note_: No Triton here.

| qtype | float16 | bfloat16 |
| - | - | - |
| Q4_0 | 1.27 | 1.09 |
| Q4_1 | 1.44 | 1.38 |
| Q5_0 | 1.06 | 1.01 |
| Q5_1 | 1.15 | 1.17 |
| Q2_K | 1.48 | 1.49 |
| Q3_K | 1.07 | 1.07 |
| Q4_K | 1.40 | 1.40 |
| Q5_K | 1.22 | 1.23 |
| Q6_K | 1.12 | 1.07 |

***

## PyTorch Q4_0 vs PyTorch other quants

This also between PyTorch implementations, no Triton. Comparing the dequantization speed vs Q4_0 (Q8_0 may be faster but I don't have it implemented). So we say Q4_0 is 1.0x speed and compare the other quants to this. Comparison done with `float32`.

| qtype | Q4_0 factor |
| - | - |
| Q4_0 | 1.0000 |
| Q4_1 | 1.3485 |
| Q5_0 | 2.8110 |
| Q5_1 | 3.1835 |
| Q2_K | 1.3208 |
| Q3_K | 1.8160 |
| Q4_K | 1.3653 |
| Q5_K | 1.9745 |
| Q6_K | 1.7487 |

***

These tests were done before my changes to eliminate the slow `to_uint32` function Q5_0 and Q5_1 use. This improved Q5_0 and Q5_1 with the PyTorch implementations by about 8-9% for `float32`.

blepping commented:
Quick'n'dirty testing with the legacy quants and Chroma Radiance (float16 output):

```plaintext
16x16 generations:

q4_0
normal: 17/18 [00:21<00:01,  1.28s/it
triton: 17/18 [00:11<00:00,  1.49it/s

q4_1
normal: 17/18 [00:22<00:01,  1.33s/it
triton: 17/18 [00:11<00:00,  1.47it/s

q5_0
normal: 17/18 [00:53<00:03,  3.15s/it
triton: 17/18 [00:12<00:00,  1.31it/s

q5_1
normal: 17/18 [00:54<00:03,  3.23s/it
triton: 17/18 [00:12<00:00,  1.33it/s

1152x1344 generation:

q5_1
normal: 15/18 [03:01<00:36, 12.11s/it (OOM)
triton: 17/18 [01:35<00:05,  5.61s/it
```

Looking pretty good, especially the 5bit quants. I also tried testing a full-sized generation with Q5_1 (bottom results). I ran out of memory twice in a row testing the non-Triton version, VRAM was around 96% used before it failed. With the Triton kernel it stayed around 80% so at least for Q5_1 it seems like using Triton uses less memory as well as being faster.

blepping committed:
Triton kernel cleanups and refactoring

blepping commented:
@city96 This needs some cleanup, but if I get it presentable would you be interested in merging these changes?

blepping committed:
Recent PyTorch versions have native support for bfloat16

blepping committed:
Fix setting ggufconfig for CLIP loaders
Show Triton status in tooltip for optimize param in advanced loader

blepping committed:
Do internal Triton dequant math in float32 by default

blepping committed:
Fix broken bitcasting in Q3_K and Q5_0
Tweak autotune configs a bit
The return of the Q8_0 Triton kernel, matches or slightly exceeds PT performance now

blepping committed:
Compatibility with Triton 3.3.1 (presumably 3.3.0 also)

city96 reviewed (commented):
@blepping I'll have to test this properly sometime, just haven't really had time lately (or access to usable hardware most of the time lol I'm writing this review on my laptop so I can't even run the code).

Most of the comments are just random nitpicks or stuff I noticed looking over the code, again, I can't run it and haven't tested the control flow so half of them could be wrong.

Also, for the final file layout, I'd probably make a new folder named `dequant`, and rename `dequant.py` to `dequant/kernel_torch.py` and the triton one to `dequant/kernel_triton.py` or something similar. Then we can have the logic for picking either in the dequant folder's init, so it picks the correct one when you import from it. Thought that might look weird on the git diff so idk.

blepping commented on dequant.py:
```diff
@@ -1,38 +1,72 @@
-
-def dequantize(data, qtype, oshape, dtype=None):
+    if qtype == gguf.GGMLQuantizationType.BF16 and HAVE_BFLOAT16:
+        return tensor.view(dtype=torch.bfloat16).reshape(oshape).to(dtype)
```
I will remove these changes and make it a separate PR. I did test though and it seemed to work fine with a model that had a bunch of BF16 tensors so I am pretty sure it couldn't be a different layout.

> also, I don't super like the `if/elif/else` being flattened

You mean you don't like:

```python
def blah():
  if condition:
    return 1
  elif other_condition:
    return 2
  else:
    return 3
```

Compared to:

```python
def blah():
  if condition:
    return 1
  if other_condition:
    return 2
  return 3
```

Linters will complain about the former version because the `elif` and `else` are redundant and it's usually considered a "code smell" but I can do it that way if you want. (This particular part isn't going to be relevant for the reviewable pull but I can make sure I follow your style preference in other places.)

blepping replied:
> Does this actually work...? ggml bf16 has a completely different data layout compared to pytorch bf16

```plaintext
>>> import numpy  as np
>>> import torch
>>> def quantize_blocks(blocks: np.ndarray) -> np.ndarray: # From gguf-py
...     n = blocks.view(np.uint32)
...     # force nan to quiet
...     n = np.where((n & 0x7fffffff) > 0x7f800000, (n & np.uint32(0xffff0000)) | np.uint32(64 << 16), n)
...     # round to nearest even
...     n = (np.uint64(n) + (0x7fff + ((n >> 16) & 1))) >> 16
...     return n.astype(np.uint16).view(np.uint8)
...
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7f841bbf9650>
>>> x = torch.randn(1000, dtype=torch.bfloat16)
>>> xnp = x.to(dtype=torch.float32).numpy()
>>> xqnp = quantize_blocks(xnp)
>>> xq = torch.tensor(xqnp)
>>> xq.dtype
torch.uint8
>>> xdq_manual = (xq.view(torch.int16).to(torch.int32) << 16).view(torch.float32)
>>> xdq_view = xq.view(dtype=torch.bfloat16).to(torch.float32)
>>> torch.equal(xdq_manual, xdq_view)
True
```

**TL;DR**: It's the same layout and just viewing is safe as long as Torch has bf16 support.

I'm not sure the GPU even needs to have BF16 support here since it's just a storage type and there's no math involved. Of course if the user has the compute dtype set to bf16 and their GPU doesn't support it then they're going to run into issues.

If you have access to a GPU without bf16 support, a simple test would just be to temporarily manifest a bf16 tensor and see if it causes problems:

```plaintext
>>> import torch
>>> torch.arange(100, dtype=torch.uint8).view(torch.bfloat16).view(torch.uint8)
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
        36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
        54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,
        72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,
        90, 91, 92, 93, 94, 95, 96, 97, 98, 99], dtype=torch.uint8)
```

blepping commented on dequant_triton.py:
```diff
@@ -0,0 +1,723 @@
+from __future__ import annotations
```
It makes the `blah: some_type | None` pattern usable with all the Python versions ComfyUI supports. Otherwise that has to be `blah: Optional[type]`. So it would be irrelevant if you don't want type annotations at all (or if you prefer the `Optional` pattern for some reason).

blepping commented on nodes.py:
```diff
@@ -164,12 +184,26 @@ class UnetLoaderGGUFAdvanced(UnetLoaderGGUF):
+                    {"default": "default"},
+                ),
                 "patch_on_device": ("BOOLEAN", {"default": False}),
+                "optimize": (
+                    ("none", "compile", "triton"),
+                    {
+                        "default": "none",
+                        "tooltip": f"Triton status: {'available' if HAVE_TRITON else 'unavailable'}\nTriton kernels: {pretty_triton_quants}",
+                    },
+                ),
```
This could be a simple `use_triton` (or whatever you want to call it) toggle. The backend config stuff supports selecting kernels individually and some kernels might not be worth using for some people. The way _I'd_ do it if it was my project is maybe have a string parameter for `triton_kernels` which might default to `*` (if you wanted the default to be using Triton when available) or you could do something like `q4_k q5_k` to just enable those two kernels or `* -q8_0` to enable everything except `q8_0`. Would also be possible to have a simple toggle in the non-advanced loader.

I tend to always choose maximum flexibility/control for my stuff but it does make stuff more complicated.

blepping commented on dequant_triton.py:
```diff
@@ -0,0 +1,723 @@
+
+    @maybestaticmethod
+    @triton.jit
+    def dequantize_kernel(
+        q_tensor_ptr,
+        out_tensor_ptr,
+        n_total_blocks,
+        DTYPE: tl.constexpr,
+        N_BLOCKS_PER_PROG: tl.constexpr,
+        CTX: tl.constexpr,
+    ) -> None:
```
I will change it to be on one line before I mark this as ready for review. (Let's leave this open so I don't forget.)

blepping commented on dequant.py:
```diff
@@ -1,38 +1,72 @@
+    triton_dequantize_functions={}
+
+
+TORCH_COMPATIBLE_QTYPES = frozenset((None, gguf.GGMLQuantizationType.F32, gguf.GGMLQuantizationType.F16))
```
I can change it if you want. Usually it's better to use non-mutable types for stuff that isn't supposed to be mutable. It's a little easier to reason about (since you know stuff won't randomly be changing) and it's usually a little more efficient also.

blepping commented on ops.py:
```diff
@@ -224,10 +229,26 @@ def forward_comfy_cast_weights(self, input, *args, **kwargs):
     """
     Dequantize weights on the fly before doing the compute
     """
+
+    _MODULE_NAMES = ("Linear", "Conv2d", "Embedding", "LayerNorm", "GroupNorm")
+
+    def __init__(self, *args, ggufconfig: Optional[GGUFConfig]=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        linear_config = ggufconfig or DEFAULT_CONFIG
+        # Ignore patch_dtype and dequant_dtype for non-Linear layers.
+        other_config = linear_config._replace(patch_dtype=None, dequant_dtype=None)
+        self.ggufconfig = linear_config
+        for module_name in self._MODULE_NAMES:
+            module = getattr(self.__class__, module_name)
+            curr_config = linear_config if module_name == "Linear" else other_config
+            setattr(self, module_name, type(module_name, (module,), {"ggufconfig": curr_config}))
+
+
```
I will change the name to `gguf_config`.

This was the best way I could think of to make sure it is present even if something else is trying to clone the class or whatever, since doing it this way make sure it's there as a default attribute. It could be done as a parameter to `__init__` for each module, but that would require more changes and I'm not positive there isn't code somewhere that might try to recreate/rebuild them and wouldn't know to pass parameters like that.

blepping commented on dequant.py:
```diff
@@ -1,38 +1,72 @@
+    dequantize_function: Optional[Callable] = None
+    dequantize_handlers: Optional[DequantizeHandlersType] = None
+
+DEFAULT_CONFIG = GGUFConfig()
```
Do you mean have it preloaded with the default PT functions and then maybe overwrite (some of) them later on?

blepping commented on dequant.py:
```diff
@@ -1,38 +1,72 @@
 import torch
 from tqdm import tqdm
 
+HAVE_BFLOAT16=hasattr(torch, "bfloat16")
```
I'll make the bfloat16 changes a separate PR.

blepping committed:
Fix compiling when Triton is enabled
Cleanups/style fixes

blepping committed:
sync

blepping commented:
@StrongerXi 

Sorry for the slow reply.

> The fact that you couldn't replicate the perf in real world models sounds strange, would you mind elaborating what you tried?

There is Torch compile support in this pull request and the benchmark tool I linked, so you can see and (presumably) replicate what I tried using that. If you can get the Torch compile approach showing equivalent performance in real world use then I'd be interested in hearing about it.

Compiling with Torch isn't always ideal/possible so it might still be useful to have an alternative approach. For example, my Linux distribution upgraded to Python 3.14. PyTorch doesn't support compiling on Python 3.14 yet (but Triton support works fine) so I'd either have to accept a large performance drop or downgrade Python if these kernels didn't exist.

There would be less justification for the maintenance burden though if it was more of a niche use case. Maybe another way to approach this would be to have ComfyUI-GGUF add some kind of support for registering optimizations/handlers and then other projects could just integrate that way and handle maintenance themselves. Not super hard to accomplish this through monkey patching the module, but that's brittle and prone to failure so a more stable way to integrate would be preferable.

blepping committed:
Update dequant_type handling

zwukong commented:
Torch compile is useful in torch2.6 ,but when i upgrade torch to 2.8 or 2.9, it takes too long , i have to bypass the compile node.

blepping committed:
Vectorized Triton kernels
Hacky LoRA caching system

blepping commented:
This isn't getting any closer to something mergeable, unfortunately. Just syncing changes.

The vectorized kernels technically enable stuff like fused matmul and I actually got fused matmul kernels working. However, it's a performance loss for DiT type models at least because they have long sequence lengths. That stuff isn't included here, possibly it would be a win for doing inference on LLMs. I plan to remove the old kernels, as far as I can see there isn't a performance cost to the new approach at least with my GPU.

This also includes a dumb, very clunky LoRA caching system. It's horrible but it actually brings GGUF LoRA performance to parity with unquantized models. I've been using it myself for some time, basically can't live without that and the Triton kernels if I am going to use a GGUF model. With Chroma and 3 or so LoRAs it brought s/it from something like 8sec to 5sec.

If anyone is still following this and wants to try it, you'll want to use a setup like this:

<img width="785" height="238" alt="image" src="https://github.com/user-attachments/assets/add3b4c8-3403-47b0-8fcf-89f3257b45a1" />

(Note: I really recommend always setting the dtypes to `target`.)

The cache is global and the change detection code I have currently doesn't seem to work so you need to make sure the clear cache node runs after any changes to the LoRAs. That node doesn't care about the parameter or input, so you could theoretically use something like the `BlehCast` node from my ComfyUI-bleh nodepack to force it to take something other than a `MODEL` as input (for example, to clear the cache after sampling). Building the LoRA cache actually isn't that slow and is faster than no caching even if you rebuild it every time you start sampling. LoRAs are cached in RAM not VRAM though so that shouldn't be necessary most of the time.

I am not sure if it's specific to the Triton stuff, but I need to run ComfyUI with pinned  memory disabled (`--disable-pinned-memory`) at least with Wan.

blepping committed:
Unbork handling for non-Triton dequant functions, make casting offloadable

blepping committed:
I thought I was so smart.
